src/effects/backends/builtin/beatgrindeffect.cpp

```diff
@@ -129,64 +111,48 @@ void BeatGrindEffect::processChannel(
         period = std::max(period, 1/8.0);
-        delay_frames = period * bufferParameters.sampleRate();
+        loopFrames = period * bufferParameters.sampleRate();
     }
-    VERIFY_OR_DEBUG_ASSERT(delay_frames > 0) {
-        delay_frames = 1;
+    VERIFY_OR_DEBUG_ASSERT(loopFrames > 0) {
+        loopFrames = 1;
     }
 
-    int delay_samples = delay_frames * bufferParameters.channelCount();
-    VERIFY_OR_DEBUG_ASSERT(delay_samples <= gs.delay_buf.size()) {
-        delay_samples = gs.delay_buf.size();
+    int loopSamples = loopFrames * bufferParameters.channelCount();
+    VERIFY_OR_DEBUG_ASSERT(loopSamples <= gs.loop.size()) {
+        loopSamples = gs.loop.size();
     }
 
-    RampingValue<CSAMPLE_GAIN> send(send_current, gs.prev_send,
-                                    bufferParameters.framesPerBuffer());
-    // Feedback the delay buffer and then add the new input.
-
-    RampingValue<CSAMPLE_GAIN> feedback(feedback_current, gs.prev_feedback,
-                                        bufferParameters.framesPerBuffer());
-    
     for (unsigned int frame = 0;
             frame < bufferParameters.samplesPerBuffer();
             frame += bufferParameters.channelCount()) {
-        CSAMPLE_GAIN send_ramped = send.getNext();
-        CSAMPLE_GAIN feedback_ramped = feedback.getNext();    
-
+  
         for (int channel = 0; channel < bufferParameters.channelCount(); channel++) {
+            // record loop
             if (gs.isRecording) {
-                if (gs.write_position >= gs.delay_buf.size()) {
+                if (gs.writeSamplePos >= gs.loop.size()) {
                     gs.isRecording = false;
                 } else {
-                    gs.delay_buf[gs.write_position] = pInput[frame + channel];
-                    gs.write_position += 1;
+                    gs.loop[gs.writeSamplePos] = pInput[frame + channel];
+                    gs.writeSamplePos += 1;
                 }
             }
-            if (gs.write_position >= delay_samples) {
-                if (gs.read_position >= delay_samples) {
-                    gs.read_position = 0;
+            // if there is nothing to output yet passtrough dry signal??
+            CSAMPLE wetValue = pInput[frame + channel];
+            // play loop
+            if (gs.writeSamplePos >= loopSamples) {
+                if (gs.readSamplePos >= loopSamples) {
+                    gs.readSamplePos = 0;
                 }
-                gs.bufferedSample[channel] = gs.delay_buf[gs.read_position];
-                gs.read_position += 1;
+                wetValue = gs.loop[gs.readSamplePos];
+                gs.readSamplePos += 1;
             }
-            pOutput[frame + channel] = SampleUtil::clampSample(
-                pInput[frame + channel] * send_ramped +
-                gs.bufferedSample[channel] * feedback_ramped);
+            pOutput[frame + channel] = 
+                pInput[frame + channel] * (1 - mixValue) +
+                wetValue * mixValue;
```

Comment:
mixValue should be ramped with a RampingValue<CSAMPLE_GAIN>. Look at the Send parameter in EchoEffect for example.